# LLM Powered Search Service - D2
11. https://applied-llms.org/ 링크는 llm 을 이용해 1년간 개발해 본 팀이 배운 것들을 정리한 아티클 입니다. 저자의 생각을 묻는 질문들에 답하기 위해서
 * 1) chunking & embedding & store
https://python.langchain.com/v0.2/docs/how_to/#text-splitters
https://python.langchain.com/v0.2/docs/integrations/text_embedding/
https://python.langchain.com/v0.2/docs/integrations/vectorstores/
 * 2) load
 * 3) retrieval
 * 4) generation
 * 5) additional (optional) component 로 구성된 RAG 코드를 작성해주세요.
* 여러 선택지가 있는 경우 최종 선택에 대한 이유도 간단히 작성해주세요.
 * (시간이 부족하다면, chunking 과 retrieval 에 더 많은 시간을 할애해주세요)
 * 예) splitter
   * 어떤 splitter 를 사용할지, 파라미터(Chunk 크기와 overlap 등)는 어떻게 정할지 고민하고 정한 이유를 주석으로 기록, 잘 모르겠다면 파라미터별 여러 가지 값을 적용한 결과를 확인해보고 어떤 기준으로 선택했는지 작성
 * embedding model
   * 어떤 embedding 모델을 고를지 고민하고 정한 이유를 주석으로 기록, 잘 모르겠다면 2개 이상을 활용 후 결과를 확인해보고 어떤 기준으로 선택했는지 작성
 * vector store
   * 어떤 Vector store 에 저장할지 고민하고 정한 이유를 주석으로 기록, 잘 모르겠다면 2개 이상을 활용 후 결과를 확인해보고 어떤 기준으로 선택했는지 작성

* 아래의 예시와 같은 프롬프트 및 정답 세트를 3개 만들고, 위에서 구성한 RAG 코드의 결과를 GPT4o 를 활용하여 채점하는 코드를 작성해주세요.
 * 예시)
   * RAG 에 대한 저자의 생각은 무엇인가?
   * RAG 와 fine tuning 에 대해 저자는 어떻게 비교하고 있나?
   * 저자가 가장 많은 부분을 할당해 설명하는 개념은 무엇인가?

* 1-11 과정을 하는 동안 작성한 코드 (프롬프트 포함) github commit


# Webbase Loader

In [ ]:
!pip install -q --upgrade langchain_community

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 8.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 974.6/974.6 kB 13.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 315.8/315.8 kB 13.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.3/125.3 kB 5.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 145.0/145.0 kB 1.2 MB/s eta 0:00:00


In [ ]:
from langchain_community.document_loaders import WebBaseLoader

In [ ]:
loader = WebBaseLoader("https://applied-llms.org/")

In [ ]:
data = loader.load()

In [ ]:
len(data)

1

In [ ]:
data

[Document(page_content='\n\n\n\n\n\n\n\n\n\n\n\n\nApplied LLMs - What We’ve Learned From A Year of Building with LLMs\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nApplied LLMs\n\n\n\n\n\n\n\n\n\n\nCourses\n\n\n\nServices\n\n\n\nJob Board\n\n\n\nTeam\n\n\n \n\n\n\n \n\n\n\n\n\n\n\n\nOn this page\n\n1 Tactical: Nuts & Bolts of Working with LLMs\n\n1.1 Prompting\n\n1.1.1 Focus on getting the most out of fundamental prompting techniques\n1.1.2 Structure your inputs and outputs\n1.1.3 Have small prompts that do one thing, and only one thing, well\n1.1.4 Craft your context tokens\n\n1.2 Information Retrieval / RAG\n\n1.2.1 RAG is only as good as the retrieved documents’ relevance, density, and detail\n1.2.2 Don’t forget keyword search; use it as a baseline and in hybrid search\n1.2.3 Prefer RAG over finetuning for new knowledge\n1.2.4 Long-context models won’t make RAG obsolete\n\n1.3 Tuning and optimizing workflows\n\n1.3.1 Step-by-step,

In [ ]:
from google.colab import drive

# Step 1: Mount Google Drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
prompt_engineering = data[0].page_content
file_path = '/content/drive/MyDrive/Colab Notebooks/applied-llms.txt'

with open(file_path, 'w') as file:
    file.write(prompt_engineering)

print(f"String saved to {file_path}")

String saved to /content/drive/MyDrive/Colab Notebooks/applied-llms.txt


# Splitter

https://python.langchain.com/v0.2/docs/how_to/#text-splitters

In [ ]:
!pip install -qU langchain-text-splitters


load & split

https://python.langchain.com/v0.2/docs/integrations/document_loaders/web_base/

https://python.langchain.com/v0.2/docs/how_to/recursive_text_splitter/

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Load example document

# https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/
# promptengineering.txt 파일은 위 링크를 webbase loader 로 load 하여 사용
with open("/content/drive/MyDrive/Colab Notebooks/applied-llms.txt") as f:
    prompt_engineering = f.read()

text_splitter = RecursiveCharacterTextSplitter(
    # Set a really small chunk size, just to show.
    chunk_size=100,
    chunk_overlap=50,
    length_function=len,
    is_separator_regex=False,
)
texts = text_splitter.create_documents([prompt_engineering])
print(texts[0])
print(texts[1])
print(texts[2])
print(texts[3])
print(texts[4])
print(texts[5])


page_content='Applied LLMs - What We’ve Learned From A Year of Building with LLMs'
page_content='Applied LLMs\n\n\n\n\n\n\n\n\n\n\nCourses\n\n\n\nServices'
page_content='Applied LLMs\n\n\n\n\n\n\n\n\n\n\nCourses\n\n\n\nServices\n\n\n\nJob Board\n\n\n\nTeam\n\n\n \n\n\n\n \n\n\n\n\n\n\n\n\nOn this page'
page_content='Job Board\n\n\n\nTeam\n\n\n \n\n\n\n \n\n\n\n\n\n\n\n\nOn this page\n\n1 Tactical: Nuts & Bolts of Working with LLMs'
page_content='1 Tactical: Nuts & Bolts of Working with LLMs\n\n1.1 Prompting'
page_content='1.1.1 Focus on getting the most out of fundamental prompting techniques'


In [ ]:
len(prompt_engineering)

94285

In [ ]:
prompt_engineering[:300]

'\n\n\n\n\n\n\n\n\n\n\n\n\nApplied LLMs - What We’ve Learned From A Year of Building with LLMs\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nApplied LLMs\n\n\n\n\n\n\n\n\n\n\nCourses\n\n\n\nServices\n\n\n\nJob Board\n\n\n\nTeam\n\n\n \n\n\n\n \n\n\n\n\n\n\n\n\nOn this page\n\n1 Tactical: Nuts & Bolts of Working with LLMs\n\n1.1 Prompting\n\n1.1.1 Focus on '

In [ ]:
len(texts)

1743

In [ ]:
type(texts)

list

In [ ]:
texts[1742]

Document(page_content='journal = {Applied LLMs},\n  year = {2024},\n  month = {Jun},\n  url = {https://applied-llms.org/}\n}')

Recursive 이외에도 다양한 splitter 를 사용해 applied-llms.txt 파일을 split해보세요.

# Embedding Models

https://python.langchain.com/v0.2/docs/how_to/embed_text/

# 임베딩 1

In [ ]:
!pip install langchain-openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 325.5/325.5 kB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 4.5 MB/s eta 0:00:00


In [ ]:
import os

# Set the API key
os.environ['OPENAI_API_KEY'] = "--"

In [ ]:
from langchain_openai import OpenAIEmbeddings

embeddings_model = OpenAIEmbeddings()

In [ ]:
type(texts[0])

langchain_core.documents.base.Document

In [ ]:
from langchain.schema import Document

# Assuming 'texts' is a list of Document objects
texts_as_strings = [doc.page_content for doc in texts]

In [ ]:
type(texts_as_strings[0])

str

In [ ]:
embeddings = embeddings_model.embed_documents(texts_as_strings)

In [ ]:
len(embeddings), len(embeddings[0])

(1743, 1536)

In [ ]:
type(embeddings)

list

# 임베딩 2

In [ ]:
!pip install langchain-huggingface

In [ ]:
from langchain_huggingface import HuggingFaceEmbeddings

embeddings_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-mpnet-base-v2")

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:11: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
from langchain_huggingface import HuggingFaceEmbeddings

embeddings_model = HuggingFaceEmbeddings()

In [ ]:
embeddings = embeddings_model.embed_documents(texts_as_strings)

In [ ]:
len(embeddings), len(embeddings[0])

(1743, 768)

## 임베딩1,2 비교 : 가장 높은 유사도 및 유사도가 높은 chuck 확인

In [ ]:
embedded_query = embeddings_model.embed_query("What are the author's thoughts on RAG?")
embedded_query[:5]

[0.013191472738981247,
 0.04903542622923851,
 0.01514696516096592,
 0.0037928614765405655,
 -0.04857663810253143]

In [ ]:
import math

def cosine_similarity(vec1, vec2):
    # Step 1: Compute the dot product
    dot_product = sum(a * b for a, b in zip(vec1, vec2))

    # Step 2: Compute the magnitudes of the vectors
    magnitude_vec1 = math.sqrt(sum(a * a for a in vec1))
    magnitude_vec2 = math.sqrt(sum(b * b for b in vec2))

    # Step 3: Compute the cosine similarity
    if magnitude_vec1 == 0 or magnitude_vec2 == 0:
        # To avoid division by zero
        return 0.0
    return dot_product / (magnitude_vec1 * magnitude_vec2)

# Example vectors
#vector1 = [1, 2, 3]
#vector2 = [4, 5, 6]

# Compute cosine similarity
#similarity = cosine_similarity(embeddings[2], embedded_query)
#print(f"Cosine Similarity: {similarity}")


In [ ]:
type(embeddings[0]), len(embeddings[0])

(list, 768)

In [ ]:
# Compute cosine similarity
co_list = []
for embedding in embeddings:
    similarity = cosine_similarity(embedding, embedded_query)
    co_list.append(similarity)

len(co_list)

1743

가장 유사도가 높은 chuck 확인

In [ ]:
def find_max_and_index(numbers):
    if not numbers:
        return None, None  # 리스트가 비어 있을 경우 None 반환

    max_value = max(numbers)
    max_index = numbers.index(max_value)
    return max_value, max_index


# 최대값과 인덱스 구하기
max_value, max_index = find_max_and_index(co_list)

print("임베딩사이즈:", len(embedded_query))
print("최대값:", max_value)
print("최대값의 인덱스:", max_index)
print("가장 유사한 문장:", texts_as_strings[max_index])


임베당사이즈: 768
최대값: 0.5132056996432574
최대값의 인덱스: 357
가장 유사한 문장: 10M tokens in size, some have begun to question the future of RAG.




---


# 임베딩 선택 결론
* 임베딩1 을 선택한다.
* cosine similarity 최대값이 0.5 로 작아, cosine similarity 최대값 0.8을 넘는 임베딩 1을 선택한다.
* 임베딩 사이즈는 임베딩 2가 더 효율적이나, 실습용 문서가 크지 않아 임베딩 사이즈에 따른 효율성은 후순위로 둔다.

## 임베딩 1
* 임베딩사이즈: 1536
* 최대값: 0.8598968161774014
* 최대값의 인덱스: 9
* 가장 유사한 문장: 1.2 Information Retrieval / RAG


## 임베딩2
* 임베딩사이즈: 768
* 최대값: 0.5132056996432574
* 최대값의 인덱스: 357
* 가장 유사한 문장: 10M tokens in size, some have begun to question the future of RAG.


---



# Vector store

In [ ]:
!pip install langchain-chroma

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 559.5/559.5 kB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.0/92.0 kB 6.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 18.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 41.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.9/59.9 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 107.0/107.0 kB 12.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 10.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 283.7/283.7 kB 35.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 47.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.6/67.6 kB 

In [ ]:
!pip install langchain_community

In [ ]:
!pip install sentence-transformers

In [ ]:
# import
from langchain_chroma import Chroma
from langchain_community.document_loaders import TextLoader
from langchain_community.embeddings.sentence_transformer import (
    SentenceTransformerEmbeddings,
)
from langchain_text_splitters import CharacterTextSplitter

# load the document and split it into chunks

from google.colab import drive
drive.mount('/content/drive')

loader = TextLoader("/content/drive/MyDrive/Colab Notebooks/applied-llms.txt")
documents = loader.load()

# split it into chunks
# chunk 사이즈, 오버랩 조정함.
text_splitter = CharacterTextSplitter(chunk_size=300, chunk_overlap=50)
docs = text_splitter.split_documents(documents)

# create the open-source embedding function
embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")

# load it into Chroma
db = Chroma.from_documents(docs, embedding_function)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


In [ ]:
len(docs)

131

In [ ]:
docs[0]

Document(page_content='Applied LLMs - What We’ve Learned From A Year of Building with LLMs\n\nApplied LLMs\n\n\nCourses\n\nServices\n\nJob Board\n\nTeam\n\n\n \n\n \n\n\nOn this page\n\n1 Tactical: Nuts & Bolts of Working with LLMs\n\n1.1 Prompting', metadata={'source': '/content/drive/MyDrive/Colab Notebooks/applied-llms.txt'})

In [ ]:
# query it
query = "What are the author's thoughts on RAG?"
results = db.similarity_search(query)

len(results)

4

In [ ]:
# print results
i=0
for result in results:
  print(i, "---", results[i].page_content)
  i+=1

0 --- 1.2.4 Long-context models won’t make RAG obsolete
With Gemini 1.5 providing context windows of up to 10M tokens in size, some have begun to question the future of RAG.
1 --- 1.2.4 Long-context models won’t make RAG obsolete
With Gemini 1.5 providing context windows of up to 10M tokens in size, some have begun to question the future of RAG.
2 --- 1.2.1 RAG is only as good as the retrieved documents’ relevance, density, and detail
1.2.2 Don’t forget keyword search; use it as a baseline and in hybrid search
1.2.3 Prefer RAG over finetuning for new knowledge
1.2.4 Long-context models won’t make RAG obsolete
3 --- 1.2.1 RAG is only as good as the retrieved documents’ relevance, density, and detail
1.2.2 Don’t forget keyword search; use it as a baseline and in hybrid search
1.2.3 Prefer RAG over finetuning for new knowledge
1.2.4 Long-context models won’t make RAG obsolete


In [ ]:
# save to disk
db2 = Chroma.from_documents(docs, embedding_function, persist_directory="./chroma_db")
results = db2.similarity_search(query)

# load from disk
db3 = Chroma(persist_directory="./chroma_db", embedding_function=embedding_function)
results = db3.similarity_search(query)
print(results[0].page_content)

1.2.4 Long-context models won’t make RAG obsolete
With Gemini 1.5 providing context windows of up to 10M tokens in size, some have begun to question the future of RAG.


In [ ]:
import chromadb

persistent_client = chromadb.PersistentClient()
collection = persistent_client.get_or_create_collection("collection_name")
collection.add(ids=["1", "2", "3"], documents=["a", "b", "c"])

langchain_chroma = Chroma(
    client=persistent_client,
    collection_name="collection_name",
    embedding_function=embedding_function,
)

print("There are", langchain_chroma._collection.count(), "in the collection")

/root/.cache/chroma/onnx_models/all-MiniLM-L6-v2/onnx.tar.gz: 100%|██████████| 79.3M/79.3M [00:05<00:00, 14.6MiB/s]


There are 3 in the collection


# Generation

https://python.langchain.com/v0.2/docs/tutorials/rag/

In [ ]:
!pip install -qU langchain-openai

In [ ]:
import getpass

from langchain_openai import ChatOpenAI

import os

# Set the API key
#os.environ['OPENAI_API_KEY'] = ""

llm = ChatOpenAI(model="gpt-3.5-turbo-0125")

In [ ]:
!pip install langchain-chroma

In [ ]:
!pip install langchainhub

In [ ]:
import bs4
from langchain import hub
from langchain_chroma import Chroma
from langchain_community.document_loaders import WebBaseLoader
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_openai import OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Load, chunk and index the contents of the blog.
loader = WebBaseLoader(
    web_paths=("https://lilianweng.github.io/posts/2023-06-23-agent/",),
    bs_kwargs=dict(
        parse_only=bs4.SoupStrainer(
            class_=("post-content", "post-title", "post-header")
        )
    ),
)
docs = loader.load()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
splits = text_splitter.split_documents(docs)
# collection_name 으로 DB 네임을 따로 주면 해결 됨.
vectorstore = Chroma.from_documents(documents=splits, collection_name="lilianweng", embedding=OpenAIEmbeddings())

# Retrieve and generate using the relevant snippets of the blog.
retriever = vectorstore.as_retriever()
prompt = hub.pull("rlm/rag-prompt")


def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)


rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)

rag_chain.invoke("What is Task Decomposition?")

'Task Decomposition is a technique used to break down complex tasks into smaller and simpler steps. This approach helps agents to better plan and execute tasks by dividing them into manageable parts. Task decomposition can be achieved through various methods such as using prompting techniques, task-specific instructions, or human input.'

In [ ]:
!pip install -U langsmith

In [ ]:
os.environ["LANGCHAIN_TRACING_V2"] = "true"

# https://smith.langchain.com/ 여기에서 간단하게 회원 가입 후 API 발급 가능 (create API KEY)
os.environ["LANGCHAIN_API_KEY"] = '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'

In [ ]:
import openai
from langsmith.wrappers import wrap_openai
from langsmith import traceable

# Auto-trace LLM calls in-context
client = wrap_openai(openai.Client())

@traceable # Auto-trace this function
def pipeline(user_input: str):
    result = client.chat.completions.create(
        messages=[{"role": "user", "content": user_input}],
        model="gpt-3.5-turbo"
    )
    return result.choices[0].message.content

pipeline("Hello, world!")
# Out:  Hello there! How can I assist you today?

'Hello! How can I assist you today?'

In [ ]:
from langsmith import Client
from langsmith.evaluation import evaluate

client = Client()

# Define dataset: these are your test cases
dataset_name = "Sample Dataset 2"
dataset = client.create_dataset(dataset_name, description="A sample dataset in LangSmith.")
client.create_examples(
    inputs=[
        {"postfix": "to LangSmithx"},
        {"postfix": "to Evaluations in LangSmith"},
    ],
    outputs=[
        {"output": "Welcome to LangSmith"},
        {"output": "Welcome to Evaluations in LangSmith"},
    ],
    dataset_id=dataset.id,
)

# Define your evaluator
def exact_match(run, example):
    return {"score": run.outputs["output"] == example.outputs["output"]}

experiment_results = evaluate(
    lambda input: "Welcome " + input['postfix'], # Your AI system goes here
    data=dataset_name, # The data to predict and grade over
    evaluators=[exact_match], # The evaluators to score the results
    experiment_prefix="sample-experiment", # The name of the experiment
    metadata={
      "version": "1.0.0",
      "revision_id": "beta"
    },
)

View the evaluation results for experiment: 'sample-experiment-5033fbd0' at:
https://smith.langchain.com/o/abfd60df-494f-5489-8346-e776a79cb346/datasets/783f0f5f-ed87-4c49-9163-fa951b235d9b/compare?selectedSessions=a4b02c13-33e4-45e6-98c9-1441a2127893




0it [00:00, ?it/s]

# Query Translation

Multi query

https://python.langchain.com/v0.2/docs/how_to/MultiQueryRetriever/

In [ ]:
# Build a sample vectorDB
from langchain_chroma import Chroma
from langchain_community.document_loaders import WebBaseLoader
from langchain_openai import OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Load blog post
loader = WebBaseLoader("https://lilianweng.github.io/posts/2023-06-23-agent/")
data = loader.load()

# Split
text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=0)
splits = text_splitter.split_documents(data)

# VectorDB
embedding = OpenAIEmbeddings()
vectordb = Chroma.from_documents(documents=splits, collection_name="lilianweng", embedding=embedding)

In [ ]:
from langchain.retrievers.multi_query import MultiQueryRetriever
from langchain_openai import ChatOpenAI

question = "What are the approaches to Task Decomposition?"
llm = ChatOpenAI(temperature=0)
retriever_from_llm = MultiQueryRetriever.from_llm(
    retriever=vectordb.as_retriever(), llm=llm
)

In [ ]:
# Set logging for the queries
# import logging

# logging.basicConfig()
# logging.getLogger("langchain.retrievers.multi_query").setLevel(logging.INFO)

In [ ]:
unique_docs = retriever_from_llm.invoke(question)
len(unique_docs)

4

In [ ]:
from typing import List

from langchain_core.output_parsers import BaseOutputParser
from langchain_core.prompts import PromptTemplate
from langchain_core.pydantic_v1 import BaseModel, Field


# Output parser will split the LLM result into a list of queries
class LineListOutputParser(BaseOutputParser[List[str]]):
    """Output parser for a list of lines."""

    def parse(self, text: str) -> List[str]:
        lines = text.strip().split("\n")
        return lines


output_parser = LineListOutputParser()

QUERY_PROMPT = PromptTemplate(
    input_variables=["question"],
    template="""You are an AI language model assistant. Your task is to generate five
    different versions of the given user question to retrieve relevant documents from a vector
    database. By generating multiple perspectives on the user question, your goal is to help
    the user overcome some of the limitations of the distance-based similarity search.
    Provide these alternative questions separated by newlines.
    Original question: {question}""",
)
llm = ChatOpenAI(temperature=0)

# Chain
llm_chain = QUERY_PROMPT | llm | output_parser

# Other inputs
question = "What are the approaches to Task Decomposition?"

In [ ]:
# Run
retriever = MultiQueryRetriever(
    retriever=vectordb.as_retriever(), llm_chain=llm_chain, parser_key="lines"
)  # "lines" is the key (attribute name) of the parsed output

# Results
unique_docs = retriever.invoke("What does the course say about regression?")
len(unique_docs)

10

# RAG fusion

In [ ]:
# os.environ['LANGCHAIN_TRACING_V2'] = 'true'
# os.environ['LANGCHAIN_ENDPOINT'] = 'https://api.smith.langchain.com'


os.environ["LANGCHAIN_TRACING_V2"] = "true"
# os.environ["LANGCHAIN_API_KEY"] = ''

In [ ]:
from langchain.prompts import ChatPromptTemplate

# RAG-Fusion: Related
template = """You are a helpful assistant that generates multiple search queries based on a single input query. \n
Generate multiple search queries related to: {question} \n
Output (4 queries):"""
prompt_rag_fusion = ChatPromptTemplate.from_template(template)

In [ ]:
prompt_rag_fusion

ChatPromptTemplate(input_variables=['question'], messages=[HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=['question'], template='You are a helpful assistant that generates multiple search queries based on a single input query. \n\nGenerate multiple search queries related to: {question} \n\nOutput (4 queries):'))])

In [ ]:
from langchain_core.output_parsers import StrOutputParser
from langchain_openai import ChatOpenAI

generate_queries = (
    prompt_rag_fusion
    | ChatOpenAI(temperature=0)
    | StrOutputParser()
    | (lambda x: x.split("\n"))
)

In [ ]:
from langchain.load import dumps, loads

def reciprocal_rank_fusion(results: list[list], k=60):
    """ Reciprocal_rank_fusion that takes multiple lists of ranked documents
        and an optional parameter k used in the RRF formula """

    # Initialize a dictionary to hold fused scores for each unique document
    # key: doc_str, value: score
    fused_scores = {}

    # Iterate through each list of ranked documents
    for docs in results:
        # Iterate through each document in the list, with its rank (position in the list)
        for rank, doc in enumerate(docs):
            # Convert the document to a string format to use as a key (assumes documents can be serialized to JSON)
            doc_str = dumps(doc)
            # If the document is not yet in the fused_scores dictionary, add it with an initial score of 0
            if doc_str not in fused_scores:
                fused_scores[doc_str] = 0
            # Retrieve the current score of the document, if any
            previous_score = fused_scores[doc_str]
            # Update the score of the document using the RRF formula: 1 / (rank + k)
            fused_scores[doc_str] += 1 / (rank + k)

    # Sort the documents based on their fused scores in descending order to get the final reranked results
    reranked_results = [
        (loads(doc), score)
        for doc, score in sorted(fused_scores.items(), key=lambda x: x[1], reverse=True)
    ]

    # Return the reranked results as a list of tuples, each containing the document and its fused score
    return reranked_results



In [ ]:
retrieval_chain_rag_fusion = generate_queries | retriever.map() | reciprocal_rank_fusion
docs = retrieval_chain_rag_fusion.invoke({"question": question})
len(docs)

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/beta_decorator.py:87: LangChainBetaWarning: The function `loads` is in beta. It is actively being worked on, so the API may change.
  warn_beta(


9

In [ ]:
docs

[(Document(page_content='Task decomposition can be done (1) by LLM with simple prompting like "Steps for XYZ.\\n1.", "What are the subgoals for achieving XYZ?", (2) by using task-specific instructions; e.g. "Write a story outline." for writing a novel, or (3) with human inputs.', metadata={'description': 'Building agents with LLM (large language model) as its core controller is a cool concept. Several proof-of-concepts demos, such as AutoGPT, GPT-Engineer and BabyAGI, serve as inspiring examples. The potentiality of LLM extends beyond generating well-written copies, stories, essays and programs; it can be framed as a powerful general problem solver.\nAgent System Overview In a LLM-powered autonomous agent system, LLM functions as the agent’s brain, complemented by several key components:', 'language': 'en', 'source': 'https://lilianweng.github.io/posts/2023-06-23-agent/', 'title': "LLM Powered Autonomous Agents | Lil'Log"}),
  0.06666666666666667),
 (Document(page_content='Fig. 1. Over

In [ ]:
from langchain_core.runnables import RunnablePassthrough
from operator import itemgetter

# RAG
template = """Answer the following question based on this context:

{context}

Question: {question}
"""

prompt = ChatPromptTemplate.from_template(template)

final_rag_chain = (
    {"context": retrieval_chain_rag_fusion,
     "question": itemgetter("question")}
    | prompt
    | llm
    | StrOutputParser()
)

final_rag_chain.invoke({"question":question})

'The approaches to Task Decomposition mentioned in the context are:\n\n1. Using LLM with simple prompting.\n2. Using task-specific instructions.\n3. Human inputs.'

# Decomposition

In [ ]:
#### INDEXING ####

# Load blog
import bs4
from langchain_community.document_loaders import WebBaseLoader
loader = WebBaseLoader(
    web_paths=("https://lilianweng.github.io/posts/2023-06-23-agent/",),
    bs_kwargs=dict(
        parse_only=bs4.SoupStrainer(
            class_=("post-content", "post-title", "post-header")
        )
    ),
)
blog_docs = loader.load()

# Split
from langchain.text_splitter import RecursiveCharacterTextSplitter
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    chunk_size=300,
    chunk_overlap=50)

# Make splits
splits = text_splitter.split_documents(blog_docs)

# Index
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import Chroma
vectorstore = Chroma.from_documents(documents=splits, collection_name="lilianweng",
                                    embedding=OpenAIEmbeddings())

retriever = vectorstore.as_retriever()

In [ ]:
from langchain.prompts import ChatPromptTemplate

# Multi Query: Different Perspectives
template = """You are an AI language model assistant. Your task is to generate five
different versions of the given user question to retrieve relevant documents from a vector
database. By generating multiple perspectives on the user question, your goal is to help
the user overcome some of the limitations of the distance-based similarity search.
Provide these alternative questions separated by newlines. Original question: {question}"""
prompt_perspectives = ChatPromptTemplate.from_template(template)

from langchain_core.output_parsers import StrOutputParser
from langchain_openai import ChatOpenAI

generate_queries = (
    prompt_perspectives
    | ChatOpenAI(temperature=0)
    | StrOutputParser()
    | (lambda x: x.split("\n"))
)

In [ ]:
from langchain.load import dumps, loads

def get_unique_union(documents: list[list]):
    """ Unique union of retrieved docs """
    # Flatten list of lists, and convert each Document to string
    flattened_docs = [dumps(doc) for sublist in documents for doc in sublist]
    # Get unique documents
    unique_docs = list(set(flattened_docs))
    # Return
    return [loads(doc) for doc in unique_docs]

# Retrieve
question = "What is task decomposition for LLM agents?"
retrieval_chain = generate_queries | retriever.map() | get_unique_union
docs = retrieval_chain.invoke({"question":question})
len(docs)

5

In [ ]:
from langchain.prompts import ChatPromptTemplate

# Decomposition
template = """You are a helpful assistant that generates multiple sub-questions related to an input question. \n
The goal is to break down the input into a set of sub-problems / sub-questions that can be answers in isolation. \n
Generate multiple search queries related to: {question} \n
Output (3 queries):"""
prompt_decomposition = ChatPromptTemplate.from_template(template)

In [ ]:
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser

# LLM
llm = ChatOpenAI(temperature=0)

# Chain
generate_queries_decomposition = ( prompt_decomposition | llm | StrOutputParser() | (lambda x: x.split("\n")))

# Run
question = "What are the main components of an LLM-powered autonomous agent system?"
questions = generate_queries_decomposition.invoke({"question":question})

In [ ]:
questions


['1. What is LLM technology and how does it work in autonomous agent systems?',
 '2. What are the specific components that make up an LLM-powered autonomous agent system?',
 '3. How do the main components of an LLM-powered autonomous agent system interact with each other to enable autonomous behavior?']

In [ ]:
# Prompt
template = """Here is the question you need to answer:

\n --- \n {question} \n --- \n

Here is any available background question + answer pairs:

\n --- \n {q_a_pairs} \n --- \n

Here is additional context relevant to the question:

\n --- \n {context} \n --- \n

Use the above context and any background question + answer pairs to answer the question: \n {question}
"""

decomposition_prompt = ChatPromptTemplate.from_template(template)

In [ ]:
from operator import itemgetter
from langchain_core.output_parsers import StrOutputParser

def format_qa_pair(question, answer):
    """Format Q and A pair"""

    formatted_string = ""
    formatted_string += f"Question: {question}\nAnswer: {answer}\n\n"
    return formatted_string.strip()

# llm
llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0)

q_a_pairs = ""
for q in questions:

    rag_chain = (
    {"context": itemgetter("question") | retriever,
     "question": itemgetter("question"),
     "q_a_pairs": itemgetter("q_a_pairs")}
    | decomposition_prompt
    | llm
    | StrOutputParser())

    answer = rag_chain.invoke({"question":q,"q_a_pairs":q_a_pairs})
    q_a_pair = format_qa_pair(q,answer)
    q_a_pairs = q_a_pairs + "\n---\n"+  q_a_pair

In [ ]:
answer

'In an LLM-powered autonomous agent system, the main components interact with each other in a cohesive manner to enable autonomous behavior. \n\n1. Planning: The agent utilizes planning to break down complex tasks into smaller, manageable subgoals. This component allows the agent to have a clear roadmap of what needs to be done and how to achieve it efficiently.\n\n2. Reflection and refinement: The agent engages in self-criticism and self-reflection over past actions, learning from mistakes and refining its strategies for future steps. By continuously improving its approach based on past experiences, the agent enhances the quality of its final results.\n\n3. Memory: The memory component enables the agent to store and retrieve information, leveraging past experiences and knowledge to make informed decisions. This interaction with memory allows the agent to learn from previous interactions and adapt its behavior accordingly, contributing to its autonomous capabilities.\n\nOverall, the in

# Routing

In [ ]:
! pip install langchain_community tiktoken langchain-openai langchainhub chromadb langchain youtube-transcript-api pytube

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 3.1 MB/s eta 0:00:00


In [ ]:
from typing import Literal

from langchain_core.prompts import ChatPromptTemplate
from langchain_core.pydantic_v1 import BaseModel, Field
from langchain_openai import ChatOpenAI

# Data model
class RouteQuery(BaseModel):
    """Route a user query to the most relevant datasource."""

    datasource: Literal["python_docs", "js_docs", "golang_docs"] = Field(
        ...,
        description="Given a user question choose which datasource would be most relevant for answering their question",
    )

# LLM with function call
llm = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0)
structured_llm = llm.with_structured_output(RouteQuery)

# Prompt
system = """You are an expert at routing a user question to the appropriate data source.

Based on the programming language the question is referring to, route it to the relevant data source."""

prompt = ChatPromptTemplate.from_messages(
    [
        ("system", system),
        ("human", "{question}"),
    ]
)

# Define router
router = prompt | structured_llm

In [ ]:
question = """Why doesn't the following code work:

from langchain_core.prompts import ChatPromptTemplate

prompt = ChatPromptTemplate.from_messages(["human", "speak in {language}"])
prompt.invoke("french")
"""

result = router.invoke({"question": question})

In [ ]:
result

RouteQuery(datasource='python_docs')

In [ ]:
result.datasource


'python_docs'

In [ ]:
def choose_route(result):
    if "python_docs" in result.datasource.lower():
        ### Logic here
        return "chain for python_docs"
    elif "js_docs" in result.datasource.lower():
        ### Logic here
        return "chain for js_docs"
    else:
        ### Logic here
        return "golang_docs"

from langchain_core.runnables import RunnableLambda

full_chain = router | RunnableLambda(choose_route)

In [ ]:
full_chain.invoke({"question": question})


'chain for python_docs'

In [ ]:
from langchain.utils.math import cosine_similarity
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import PromptTemplate
from langchain_core.runnables import RunnableLambda, RunnablePassthrough
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

# Two prompts
physics_template = """You are a very smart physics professor. \
You are great at answering questions about physics in a concise and easy to understand manner. \
When you don't know the answer to a question you admit that you don't know.

Here is a question:
{query}"""

math_template = """You are a very good mathematician. You are great at answering math questions. \
You are so good because you are able to break down hard problems into their component parts, \
answer the component parts, and then put them together to answer the broader question.

Here is a question:
{query}"""

# Embed prompts
embeddings = OpenAIEmbeddings()
prompt_templates = [physics_template, math_template]
prompt_embeddings = embeddings.embed_documents(prompt_templates)

# Route question to prompt
def prompt_router(input):
    # Embed question
    query_embedding = embeddings.embed_query(input["query"])
    # Compute similarity
    similarity = cosine_similarity([query_embedding], prompt_embeddings)[0]
    most_similar = prompt_templates[similarity.argmax()]
    # Chosen prompt
    print("Using MATH" if most_similar == math_template else "Using PHYSICS")
    return PromptTemplate.from_template(most_similar)


chain = (
    {"query": RunnablePassthrough()}
    | RunnableLambda(prompt_router)
    | ChatOpenAI()
    | StrOutputParser()
)

print(chain.invoke("What's a black hole"))

Using PHYSICS
A black hole is a region in space where the gravitational pull is so strong that nothing, not even light, can escape from it. This happens when a massive star collapses in on itself, creating a singularity with infinite density at its center. The boundary surrounding the black hole is called the event horizon. Black holes can come in different sizes, from stellar-mass black holes to supermassive black holes found at the center of galaxies.


# Query Structuring

In [ ]:
from langchain_community.document_loaders import YoutubeLoader

docs = YoutubeLoader.from_youtube_url(
    "https://www.youtube.com/watch?v=pbAd8O1Lvm4", add_video_info=True
).load()

docs[0].metadata

{'source': 'pbAd8O1Lvm4',
 'title': 'Self-reflective RAG with LangGraph: Self-RAG and CRAG',
 'description': 'Unknown',
 'view_count': 18584,
 'thumbnail_url': 'https://i.ytimg.com/vi/pbAd8O1Lvm4/hq720.jpg',
 'publish_date': '2024-02-07 00:00:00',
 'length': 1058,
 'author': 'LangChain'}

In [ ]:
import datetime
from typing import Literal, Optional, Tuple
from langchain_core.pydantic_v1 import BaseModel, Field

class TutorialSearch(BaseModel):
    """Search over a database of tutorial videos about a software library."""

    content_search: str = Field(
        ...,
        description="Similarity search query applied to video transcripts.",
    )
    title_search: str = Field(
        ...,
        description=(
            "Alternate version of the content search query to apply to video titles. "
            "Should be succinct and only include key words that could be in a video "
            "title."
        ),
    )
    min_view_count: Optional[int] = Field(
        None,
        description="Minimum view count filter, inclusive. Only use if explicitly specified.",
    )
    max_view_count: Optional[int] = Field(
        None,
        description="Maximum view count filter, exclusive. Only use if explicitly specified.",
    )
    earliest_publish_date: Optional[datetime.date] = Field(
        None,
        description="Earliest publish date filter, inclusive. Only use if explicitly specified.",
    )
    latest_publish_date: Optional[datetime.date] = Field(
        None,
        description="Latest publish date filter, exclusive. Only use if explicitly specified.",
    )
    min_length_sec: Optional[int] = Field(
        None,
        description="Minimum video length in seconds, inclusive. Only use if explicitly specified.",
    )
    max_length_sec: Optional[int] = Field(
        None,
        description="Maximum video length in seconds, exclusive. Only use if explicitly specified.",
    )

    def pretty_print(self) -> None:
        for field in self.__fields__:
            if getattr(self, field) is not None and getattr(self, field) != getattr(
                self.__fields__[field], "default", None
            ):
                print(f"{field}: {getattr(self, field)}")

In [ ]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

system = """You are an expert at converting user questions into database queries. \
You have access to a database of tutorial videos about a software library for building LLM-powered applications. \
Given a question, return a database query optimized to retrieve the most relevant results.

If there are acronyms or words you are not familiar with, do not try to rephrase them."""
prompt = ChatPromptTemplate.from_messages(
    [
        ("system", system),
        ("human", "{question}"),
    ]
)
# llm = ChatOpenAI(model="gpt-4o", temperature=0)
llm = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0)
structured_llm = llm.with_structured_output(TutorialSearch)
query_analyzer = prompt | structured_llm

In [ ]:
query_analyzer.invoke({"question": "rag from scratch"}).pretty_print()


content_search: rag from scratch
title_search: rag from scratch


In [ ]:
# model gpt-4o 를 할 때와 gpt-3.5-turbo-0125 를 할 때 결과 비교
query_analyzer.invoke({"question": "rag from scratch"})

TutorialSearch(content_search='rag from scratch', title_search='rag from scratch', min_view_count=None, max_view_count=None, earliest_publish_date=None, latest_publish_date=None, min_length_sec=None, max_length_sec=None)

In [ ]:
query_analyzer.invoke(
    {"question": "videos on chat langchain published in 2023"}
).pretty_print()

content_search: chat langchain
title_search: chat langchain
earliest_publish_date: 2023-01-01
latest_publish_date: 2024-01-01


In [ ]:
query_analyzer.invoke(
    {"question": "videos that are focused on the topic of chat langchain that are published before 2024"}
).pretty_print()

content_search: chat langchain
title_search: chat langchain
earliest_publish_date: 2024-01-01


In [ ]:
query_analyzer.invoke(
    {
        "question": "how to use multi-modal models in an agent, only videos under 5 minutes"
    }
).pretty_print()

content_search: multi-modal models in agent
title_search: multi-modal models agent
max_length_sec: 300


# Indexing - Multi representation

In [ ]:
query_analyzer.invoke(
    {"question": "videos on chat langchain published in 2024"}
).pretty_print()

content_search: chat langchain
title_search: chat langchain
earliest_publish_date: 2024-01-01
latest_publish_date: 2025-01-01


In [ ]:
# import os
# os.environ['LANGCHAIN_TRACING_V2'] = 'true'
# os.environ['LANGCHAIN_ENDPOINT'] = 'https://api.smith.langchain.com'
# os.environ['LANGCHAIN_API_KEY'] = <your-api-key>
# os.environ['OPENAI_API_KEY'] = ""


In [ ]:
from langchain_community.document_loaders import WebBaseLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter

loader = WebBaseLoader("https://lilianweng.github.io/posts/2023-06-23-agent/")
docs = loader.load()

loader = WebBaseLoader("https://lilianweng.github.io/posts/2024-02-05-human-data-quality/")
docs.extend(loader.load())

In [ ]:
import uuid

from langchain_core.documents import Document
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

chain = (
    {"doc": lambda x: x.page_content}
    | ChatPromptTemplate.from_template("Summarize the following document:\n\n{doc}")
    | ChatOpenAI(model="gpt-3.5-turbo",max_retries=0)
    | StrOutputParser()
)

summaries = chain.batch(docs, {"max_concurrency": 5})

In [ ]:
summaries

['The document discusses the concept of building autonomous agents powered by Large Language Models (LLMs), with a focus on planning, memory, and tool use components. It provides details on how LLMs can decompose tasks, engage in self-reflection, utilize different types of memory, and access external tools through APIs. The document also highlights case studies like scientific discovery agents and generative agents simulation, along with challenges such as finite context length, natural language interface reliability, and long-term planning difficulties. The author also provides citations and references for further reading.',
 'The document discusses the importance of high-quality human data for training deep learning models. It covers topics such as the wisdom of the crowd, rater agreement, and two paradigms for data annotation. Methods for improving data quality, such as influence functions and noisy cross-validation, are also explored. The document provides insights on how to identi

In [ ]:
from langchain.storage import InMemoryByteStore
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import Chroma
from langchain.retrievers.multi_vector import MultiVectorRetriever

# The vectorstore to use to index the child chunks
vectorstore = Chroma(collection_name="summaries",
                     embedding_function=OpenAIEmbeddings())

# The storage layer for the parent documents
store = InMemoryByteStore()
id_key = "doc_id"

# The retriever:
retriever = MultiVectorRetriever(
    vectorstore=vectorstore,
    byte_store=store,
    id_key=id_key,
)
doc_ids = [str(uuid.uuid4()) for _ in docs]

# Docs linked to summaries
summary_docs = [
    Document(page_content=s, metadata={id_key: doc_ids[i]})
    for i, s in enumerate(summaries)
]

# Add
retriever.vectorstore.add_documents(summary_docs)
retriever.docstore.mset(list(zip(doc_ids, docs)))

In [ ]:
query = "Memory in agents"
sub_docs = vectorstore.similarity_search(query,k=1)
sub_docs[0]

Document(page_content='The document discusses the concept of building autonomous agents powered by Large Language Models (LLMs), with a focus on planning, memory, and tool use components. It provides details on how LLMs can decompose tasks, engage in self-reflection, utilize different types of memory, and access external tools through APIs. The document also highlights case studies like scientific discovery agents and generative agents simulation, along with challenges such as finite context length, natural language interface reliability, and long-term planning difficulties. The author also provides citations and references for further reading.', metadata={'doc_id': '4b944cba-db84-4a45-8033-7c189015d5b1'})

In [ ]:
# entire article
retrieved_docs = retriever.get_relevant_documents(query,n_results=1)
retrieved_docs[0].page_content[0:500]

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(


"\n\n\n\n\n\nLLM Powered Autonomous Agents | Lil'Log\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nLil'Log\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nPosts\n\n\n\n\nArchive\n\n\n\n\nSearch\n\n\n\n\nTags\n\n\n\n\nFAQ\n\n\n\n\nemojisearch.app\n\n\n\n\n\n\n\n\n\n      LLM Powered Autonomous Agents\n    \nDate: June 23, 2023  |  Estimated Reading Time: 31 min  |  Author: Lilian Weng\n\n\n \n\n\nTable of Contents\n\n\n\nAgent System Overview\n\nComponent One: Planning\n\nTask Decomposition\n\nSelf-Reflection\n\n\nComponent Two: Memory\n\nTypes of Memory\n\nMaximum Inner Product Search (MIPS)\n\n"